src/boot_O2/loadfragment.c

```diff
@@ -1,8 +1,187 @@
+/**
+ * @file loadfragment.c
+ *
+ * Functions used to process and relocate overlays
+ *
+ * Note:
+ *  These are completly unused in favor of the functions in `loadfragment2.c`.
+ *
+ *  The main difference between them seems to be the lack of vRamEnd arguments here.
+ *  Instead they are calculated on the fly.
+ */
 #include "global.h"
 #include "system_malloc.h"
 
-#pragma GLOBAL_ASM("asm/non_matchings/boot/loadfragment/Load_Relocate.s")
+s32 gLoadLogSeverity = 2;
 
-#pragma GLOBAL_ASM("asm/non_matchings/boot/loadfragment/Load_LoadOverlay.s")
+void Load_Relocate(void* allocatedVRamAddr, OverlayRelocationSection* ovl, uintptr_t vRamStart) {
+    u32 sections[4];
+    u32* relocDataP;
+    u32 reloc;
+    uintptr_t relocatedAddress;
+    u32 i;
+    u32* luiInstRef;
+    uintptr_t allocu32 = (uintptr_t)allocatedVRamAddr;
+    u32* regValP;
+    u32* luiRefs[32];
+    u32 luiVals[32];
+    u32 isLoNeg;
 
-#pragma GLOBAL_ASM("asm/non_matchings/boot/loadfragment/Load_AllocateAndLoad.s")
+    if (gLoadLogSeverity >= 3) {}
+
+    sections[0] = 0;
+    sections[1] = allocu32;
+    sections[2] = allocu32 + ovl->textSize;
+    sections[3] = sections[2] + ovl->dataSize;
+
+    for (i = 0; i < ovl->nRelocations; i++) {
+        reloc = ovl->relocations[i];
+        relocDataP = (u32*)(sections[reloc >> 0x1E] + (reloc & 0xFFFFFF));
+
+        switch (reloc & 0x3F000000) {
+            case 0x2000000:
+                // R_MIPS_32
+                // Handles 32-bit address relocation, used for things such as jump tables and pointers in data.
+
+                // Check address is valid for relocation
+                if ((*relocDataP & 0xF000000) == 0) {
+                    *relocDataP = *relocDataP - vRamStart + allocu32;
+                } else if (gLoadLogSeverity >= 3) {
+                }
+                break;
+
+            case 0x4000000:
+                // R_MIPS_26
+                // Handles 26-bit address relocation, used for jumps and jals.
+
+                *relocDataP =
+                    (*relocDataP & 0xFC000000) |
+                    ((((((*relocDataP & 0x3FFFFFF) << 2) | 0x80000000) - vRamStart + allocu32) & 0xFFFFFFF) >> 2);
+                break;
+
+            case 0x5000000:
+                // R_MIPS_HI16
+                // Handles relocation for a hi/lo pair, part 1.
+                // Store the reference to the LUI instruction, and update it in the R_MIPS_LO16 section.
+
+                luiRefs[(*relocDataP >> 0x10) & 0x1F] = relocDataP;
+                luiVals[(*relocDataP >> 0x10) & 0x1F] = *relocDataP;
```

Comment:
`>> 0x10 & 0x1F` extracts the `rt` field, again not sure if that should be made more clear

Reply:
Yeah, not sure either if that needs to be clearer. Thoughts from others?

Reply:
I think a comment about it being the destination register would be reasonable. Either that or explain actually dissecting the instruction into `opcode | rt | imm` and so on for each.

Reply:
Does this seem fine? Explaining the process a little more in detail in the header comments of each case
```
            case R_MIPS_32 << 24:
                // Handles 32-bit address relocation, used for things such as jump tables and pointers in data.
                // Just relocate the full address

            [...]

            case R_MIPS_26 << 24:
                // Handles 26-bit address relocation, used for jumps and jals.
                // Extract the address from the target field of the J-type MIPS instruction.
                // Relocate the address and update the instruction.

               [...]

            case R_MIPS_HI16 << 24:
                // Handles relocation for a hi/lo pair, part 1.
                // Store the reference to the LUI instruction (hi) using the `rt` register of the instruction.
                // This will be updated later in the `R_MIPS_LO16` section.

               [...]

            case R_MIPS_LO16 << 24:
                // Handles relocation for a hi/lo pair, part 2.
                // Grab the stored LUI (hi) from the `R_MIPS_HI16` section using the `rs` register of the instruction.
                // The full address is calculated, relocated, and then used to update both the LUI and lo parts.
                // If the lo part is negative, add 1 to the LUI.

               [...]
```

Reply:
I'm also debating if I should add a
`#define RELOCATE_ADDRESS(address, vRamStart, allocu32) ((address) - (vramStart) + (allocu32))`
macro just to make it clear what is actually getting relocated, but not sure.

Reply:
I like the idea of this macro

Reply:
Added. Let me know what you think.